### PR TITLE
AVDevice: version 1.3.1

### PR DESCRIPTION
**Pull request by onkelandy**
PEP8 optimized code
updated plugin.yaml and README
updated reset function for Denon receivers when no correct response received
updated automatic optimization of send command list
added debug verbose levels
full zone 2 and 3 control possible
added a lot of new commands to Denon text file

### CODE REVIEW

**Comment by onkelandy** [comment]
Newly created request, able to merge...

Concerning verbose level: how can I do that using logging.yaml? I want to have different levels of debug, not only info/debug..

**Comment by bmxp** [comment]
The logging level like logging.DEBUG or logging.CRITICAL are essentially just integers. You can define and use own levels: like DEBUGVERBOSE = 9 or VERYDEBUGVERBOSE = 1

**Comment by onkelandy** [comment]
Didn't really get it.. How do I use that?
self.logger.verbose1("Blabla")
and in logging.yaml I assign level: verbose1?

The link here to the best practice https://www.smarthomeng.de/dev/logging.html#id6 actually doesn't lead anywhere :(

**Comment by bmxp** [comment]
Link to best practices can be found [here](https://github.com/smarthomeNG/smarthome/wiki/Logging----Best-Practices).

Some more info [here](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial)

For the level:
```python
import logging
DEBUGVERBOSE = logging.DEBUG - 1
logger.log(DEBUGVERBOSE , 'this is my verbose log message')
```

should work (not tested however)

**Comment by msinn** [comment]
The official Python documentation says:
> Defining your own levels is possible, but should not be necessary, as the existing levels have been chosen on the basis of practical experience. 

Two sentences later they state that it's
> a very bad idea to define custom levels if you are developing a library.

**Comment by ohinckel** [comment]
Even if it is bad practice, it is the best if you require custom log levels. I don't see any better solution for it.

And - if one requires this logging - it can be configured explicitely for the plugin instead globally to avoid confusion with different plugin/libraries/etc. using same log levels and producing log messages.

File: avdevice/plugin.yaml [hunk 1]
```diff
@@ -3,23 +3,162 @@ plugin:
...
+        type: int
+        default: 0
+        description:
+            de: "Ein Wert zwischen 0 und 3, um das Verboselevel des Debugloggers zu definieren. Je höher, desto mehr Nachrichten werden geloggt."
+            en: "Value between 0 and 3 to define the verbose level of the debug logger. The higher the value, the more debug messages."
+  
 item_attributes:
-    # Definition of item attributes defined by this plugin
+    avdevice_zone[0-4]:
+        type: name[a-z]
```
Comment (ohinckel) [inline avdevice/plugin.yaml]: @msinn, what about the "regex-matching-attribute-name" feature (as mentioned in #81)? Could it be useful for numbered attribute names with same settings?

Further what is the type `name[a-z]`?

**Comment by onkelandy** [comment]
Any conclusion concerning custom log levels?
Might try the logger.log stuff then...

**Comment by bmxp** [comment]
Go ahead and try if it helps. 

> Developing a library 
IMHO means that you should not create other loglevels when e.g. write another library for PyPI. It is ok for this internal little project I think.

**Comment by onkelandy** [comment]
oki, updated everything and creating new PR

**Pull request closed without merging**